test/e2e/volumes.go

```diff
@@ -0,0 +1,360 @@
+*/
+
+/*
+ * This test checks that various VolumeSources are working. For each volume
+ * type it creates a server pod, exporting simple 'index.html' file.
+ * Then it uses appropriate VolumeSource to import this file into a client pod
+ * and checks that the pod can see the file. It does so by importing the file
+ * into web server root and loadind the index.html from it.
+ *
+ * These tests work only when privileged containers are allowed, exporting
+ * various filesystems (NFS, GlusterFS, ...) usually needs some mounting or
+ * other privileged magic in the server pod.
+ *
+ * Note that the server containers are for testing purposes only and should not
+ * be used in production.
+ */
+
+package e2e
+
+import (
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// Configuration of one tests. The test consist of:
+// - server pod - runs serverImage, exports ports[]
+// - client pod - does not need any special configuration
+type VolumeTestConfig struct {
+	namespace string
+	// Prefix of all pods. Typically the test name.
+	prefix string
+	// Name of container image for the server pod.
+	serverImage string
+	// Ports to export from the server pod. TCP only.
+	serverPorts []int
+}
+
+// Starts a container specified by config.containerImage and exports all
+// configured ports from it. The returned pod should be used to get the server
+// IP address and create appropriate VolumeSource.
+// The function may return nil - the test should be skipped in this case.
+func startVolumeServer(client *client.Client, config VolumeTestConfig) *api.Pod {
+	podClient := client.Pods(config.namespace)
+
+	portCount := len(config.serverPorts)
+	serverPodPorts := make([]api.ContainerPort, portCount)
+
+	for i := 0; i < portCount; i++ {
+		portName := fmt.Sprintf("%s-%d", config.prefix, i)
+
+		serverPodPorts[i] = api.ContainerPort{
+			Name:          portName,
+			ContainerPort: config.serverPorts[i],
+			Protocol:      api.ProtocolTCP,
+		}
+	}
+
+	By(fmt.Sprint("creating ", config.prefix, " server pod"))
+	serverPod := &api.Pod{
+		TypeMeta: api.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1beta3",
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name: config.prefix + "-server",
+			Labels: map[string]string{
+				"role": config.prefix + "-server",
+			},
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:       config.prefix + "-server",
+					Image:      config.serverImage,
+					Privileged: true,
+					Ports:      serverPodPorts,
+				},
+			},
+		},
+	}
+	if _, err := podClient.Create(serverPod); err != nil {
+		// Skip the test when 'privileged' containers are not allowed
+		if strings.Contains(err.Error(), "spec.containers[0].privileged: forbidden 'true'") {
+			By(fmt.Sprint("Skipping ", config.prefix, " test, which is supported only on on cluster with --allow_privileged=True"))
```

Comment:
I have no strong opinion here, I took this pattern from other tests, e.g. networking.go:
```
                if testContext.Provider == "vagrant" {
                        By("Skipping test which is broken for vagrant (See https://github.com/GoogleCloudPlatform/kubernetes/issues/3580)")
                        return
                }
```
or pod.go:
```
                if !providerIs("gce", "aws") {
                        By(fmt.Sprintf("Skipping PD test, which is only supported for providers gce & aws (not %s)",
                                testContext.Provider))
                        return
                }
```

By default, cluster/kube-up.sh (used by hack/e2e.go --up --push --test) installs Kubernetes with --allow-privileged=False. What about switching the default to True?